lib/plugins/oauth2TokenParser.js

```diff
@@ -0,0 +1,90 @@
+'use strict';
+
+var errors = require('restify-errors');
+var querystring = require('qs');
+/*
+
+  Parses the header for the authorization: bearer
+
+*/
+function parseHeader(req) {
+
+    if (req.headers && req.headers.authorization) {
+
+        var credentialsIndex = 1;
+        var parts = req.headers.authorization.split(' ');
+        var partsExpectedLength = 2;
+        var schemeIndex = 0;
+
+        if (parts.length === partsExpectedLength) {
+
+            var credentials = parts[credentialsIndex];
+            var scheme = parts[schemeIndex];
+
+            if (/^Bearer$/i.test(scheme)) {
+                return credentials;
+            }
+
+        }
+
+    }
+
+    return null;
+
+}
+
+
+/**
+ * Returns a plugin that will parse the client's request for an OAUTH2
+   access token
+ *
+ * Subsequent handlers will see `req.oauth2`, which looks like:
+ *
+ * {
+ *   oauth2: {
+        accessToken: 'mF_9.B5f-4.1JqM&p=q'
+    }
+ * }
+ *
+ *
+ * @public
+ * @function oauth2TokenParser
+ * @throws   {InvalidArgumentError}
+ * @param    {Object} options an options object
+ * @returns  {Function}
+ */
+function oauth2TokenParser(options) {
+
+    function parseOauth2Token(req, res, next) {
+        req.oauth2 = { accessToken: null};
+
+        var tokenFromHeader = parseHeader(req);
+
+        if (tokenFromHeader) {
+
+            req.oauth2.accessToken = tokenFromHeader;
+        }
+
+        var params = querystring.parse(req.body);
```

Comment:
Just noticed this bit now that I look at it again this morning - what if user has declared bodyParser() middleware prior to oauthtoken parser? This would not work, since `req.body` is an object, right? This is complicated by the fact that if you do use bodyParser, original post body will become req._body, so it's not consistent. 

I think we should just say if you want to parse oauth tokens out of the body, you must use the bodyParser first. That way, this handler is just in charge of putting on the `oauth2` namespace plus any format validation, if such a thing needs to happen in the future. 

Thus, something like:

```
server.use(plugins.bodyParser());
server.use(plugins.oauth2TokenParser());
```

But if you choose not to use bodyParser, we'll still parse it out of the header. Otherwise, as it is currently is, seems like we're doing extra work parsing the body twice, and in some scenarios (if the postbody is a diff format) this handler would probably be completely hosed. I now realize that's probably what @baabel was trying to account for with the original try/catch - sorry for the churn here. :( 

TL,DR;, let's assume req.body is already parsed here, and if it's not, then don't set the token. That way you can choose to use the token parser to get header parsing by default, but if you add a bodyparser in front you'll get token parsing from the the post body too. 

@micahr @baabel does that sound reasonable?